README
lsid-perl version 1.1.4
=======================

* ABOUT

This module implements an object-oriented interface for building clients
and servers that resolve LSIDs and perform metadata queries on LSIDs.


* What's new in this release

This release contains updates and many bug fixes. Specific issues
resolved:

	- Fixed bug in Perl use statements
	- Minor code style cleanups

This release is based on the specification submitted to the Object
Management Group (OMG) and is not compatible with all previous releases.
Clients and authorities based on previous versions should still
interoperate with minimal changes to each. Most changes involve port
types and the addition of a Services WSDL document that describes how to
invoke the authority. Please see the examples/ for more information.


* IMPORTANT: You need SOAP::Lite 0.69 installed for this release. You should 
			 also install File::Temp from CPAN as the File::Temp included
			 with Perl is not sufficient for this module.




* MODIFICATIONS TO AUTHORITIES

This release modifies how ports are added to authorities. The OLD method
of adding ports should still work. However the new method is much cleaner:

NOTE: $authority is a LS::Service::Authority object reference

$authority->addMetadataPort(
	serviceName=> 'myService',
	endpoint=> "$location/authority/metadata",
	protocol=> $LS::Authority::WSDL::Constants::Protocols::SOAP,
	);

$authority->addDataPort(
	serviceName=> 'myService',
	endpoint=> "$location/authority/data",
	protocol=> $LS::Authority::WSDL::Constants::Protocols::SOAP,
	);

Notice the 'serviceName' and 'protocol' parameters. We recommend 
that you use different service names for each protocol. For example, 
if you have SOAP and HTTP ports for metadata, add each port to 
a service named for its respective protocol (ie. authoritySOAP 
or authorityHTTP).




FEATURES



* LSID Assigning Service

This release provides a client and the necessary framework for
implementing an LSID Assigning Service. See examples/assigning for more
information about how to use the framework and client.


* Foreign Authority Notification service

The LSID Authority framework has been updated to support FAN. See
rfcfan.txt for more information about FAN.


* Authority Framework

The Authority framework has been moved to LS::Service and underwent
significant changes. In many cases, crating a SOAP based authority
has become easier. See the examples/ subdirectory for two authority
examples.



* Caching

The cache modules have been updated to return LS::Service::Response
objects containing IO::File objects in the response.

The client stack now caches all metadata, data and WSDL queries. The
default path is stored in /tmp/lsid-client and can be change through
then evironment variable LSID_CACHE_ROOT.



DEPENDENCIES

lsid-perl requires these other modules and libraries:

SOAP-Lite-0.69
URI-1.21
libwww-perl-5.53
Net::FTP 2.65
Net-DNS-0.28
XML-XPath-1.12
MIME-Base64-2.12
File::Temp 0.12 (Install from CPAN, File::Temp in the Perl distribution is not enough)
RDF-Core-0.30
MIME-tools >6.100 (Needed for Apache AXIS compatibility)

This module is known to work with the versions given, but later and
earlier ones may also work.



INSTALLATION

lsid-perl can be installed like any Perl module found on CPAN. In short:

On Unix or Linux:

   perl Makefile.PL
   make
   make test
   make install
   
On Windows:

Download the nmake utility from
ftp://ftp.microsoft.com/Softlib/MSLFILES/nmake15.exe, then

   perl Makefile.PL
   nmake
   nmake test
   nmake install

OR

   Copy LS.pm and the LS folder into your Perl's lib directory. You can
   find the lib directory by typing

    perl -e 'print "@INC"'

For more information on installing Perl modules, see
http://search.cpan.org/author/JHI/perl-5.8.0/pod/perlmodinstall.pod


COPYRIGHT AND LICENSE

Copyright (c) 2002,2003 IBM Corporation 
All rights reserved.   This program and the accompanying materials
are made available under the terms of the Common Public License v1.0
which accompanies this distribution, and is available at
http://www.opensource.org/licenses/cpl.php


Changes
Revision history for Perl extension lsid-perl.

1.1.6
	- Again, I encountered an error with the module that was uploaded to cpan.
	  Nothing changed, except that third time is a charm! Sorry.

1.1.5
	- Error occured while uploading cpan module ... Have to up the version number although nothing new has been added. Sorry.


1.1.4
	- Fixed bug in Perl use statements
	- Minor code style cleanups

1.1.3
	- Updated the WSDL Object to be aware of XML Namespaces
	- Added default service name processing for LS::Authority::WSDL::Services
	- Fixed mispelling in MANIFEST file
	- Converted LS::Client::BasicResolver to use a simple die with error details
	- Converted to simple die with error details in LS::Authority, LS::Client::SOAP and LS::Client::HTTP. Full conversion to Error.pm is pending
	- Added several exceptions in LS.pm
	- Included new Perl module dependency: Error.pm for exceptions
	- Bug fixes in the LS::Service layer, LS::Client layer and base LS layer
	- Split LS::Service::* into individual files
	- 'use base' where possible
	- Fixed the test suite
	- Continued to cleanup LS::Authority::WSDL
	- Removed dependency on SOAP::MIME (had to work around a SOAP::Lite 0.60 bug)
	- Updated WSDL module to explicitly set XML prefix->URI mappings

1.1.2
	- Bug fixes to LS/Authority/WSDL/Service.pm

1.1.1
	- Bug fixes in examples/client/lsid_client.pl
	- Bug fixes to LS/Authority.pm
	- Added LS::RDF::ComplexDocument
	- Added LS::RDF::Bag
	- Added LS::RDF::Metadata
	- Made the Format returned in an LS::Service::Response do something to HTTP headers
	- Documentation update in all Perl modules
	- Migration from ->isa to UNIVERSAL::isa to avoid dreferencing undefined objects
	- Added LS::Base as the base class for all LS objects
	- Added error handline methods to LS::Base
	- Added stack trace information to LS::Base
	- Migrated all error handline to LS::Base routines
	- Added LS::Config that will parse the file examples/config/lsid-client.xml
	- Cleaned up more code in LS::Authority and LS::Locator
	- Added LS::Service::AdvancedDataService which allows regular expression matching in the namespace lookup
	- Added proxy authority example.
	- Added convenience methods for adding ports to an LS::Service::Authority
	- Added Web Resolver example
	- Added Authority designer example
	- Added test suite use 'make test' before 'make install'
	- Fixed numerous bugs in the WSDL parser
	- Created a LS::Authority::WSDL::Constants
	
1.1.0
	- Move to new version

1.0.3
	- Creation of a Client library
	- Creation of a basic client resolver
	- Many small fixes to the Authority framework
	- Small fixes to the example authorities
	- Rewrite of the example client
	- Fixed - [ Bug #3985 ] uri_unescape not called in LS::HTTP::Service::do_data
	- Rewrite of the cache manager, supports large files
	- Vanilla HTTP transport supports large file downloads
	- Initial implementation of LSID Assigning service framework and client
	- Seperation of LS::SOAP::Fault class to its own package
	- Removal of LS::Service::Fault::SOAP class (it was redundant)
	- Standardization of return values (LS::Service::Response or LS::Service::Fault)
	- LS::Authority::WSDL package and classes have been enhanced and made more general
	- Implmentation of Foreign Authority Notification service
	- Updated example authorities to work with new WSDL objects
	- Added LSID Assigning Service example and template (see examples/tools for the template)

1.0.2
	- Faults now percolate through the stack
 	- getServices is automatically populated with the correct endpoints
 	- Fault codes are now updated to the specification
 	- Namespaces allow full range of methods defined in the specification
 	- LSIDs are now completely case sensitive from the namespace:object:revsion portion
 	- WSDL document generation is now simpler
 	- LSID port types are now fully specified 
 	- Many service cleanups and descriptive error messages 
 	- And many more fixes...

1.0.1
	- Moved to new specification based on the OMG
	- SOAP with MIME attachments
	- HTTP Authorities with Faults
	- SOAP Authorities with Faults
	- Moved LS::Authority::Service::DataService to LS::Service
	- LS::SOAP::Service, LS::HTTP::Service are implementations of LS::Service
	- Touched up cache manager code pending rewrite
	- Added httpd.conf examples for Apache authorities
	- Some methods have been deprecated (getAuthorityVersion, getKnownURIs, as well as object methods)

0.07
	- Added cache manager for all LSID related information
	- Added HTTP Basic Authentication processing
	- Created authority framework

0.06
	- dies from RDF::Core::Model::Parser->parse are now caught
	- updated sample-client.pl
	
0.05
	- LSID strings are now validated in LS::ID->new
	- Removed resolveMetaDataQuery SOAP calls and perl interfaces
	- LS::Resource now uses the getMetaData call, and RDF::Core for processing the metadata
	- The APIs to LS::Resource and LS::Authority have changed!  See their perldoc for details.

0.03
	- Added support for SOAP data port bindings
	- Added support for getMetaData call
	- Added support for expires header
	- The APIs to LS::Resource and LS::Authority::WSDL::Simple have changed!
	  Some methods have been deprecated.  See their perldoc for details.


0.01  Mon Feb  3 14:38:15 2003
	- original version; created by h2xs 1.22 with options
		-AX -n LS -b 5.6.0

		
$Id: Changes,v 1.27 2006/07/06 17:47:08 evanchsa-oss Exp $

lib/LS.pm
# =====================================================================
# Copyright (c) 2002,2003 IBM Corporation
# All rights reserved.  This program and the accompanying materials
# are made available under the terms of the Common Public License v1.0
# which accompanies this distribution, and is available at
# http://www.opensource.org/licenses/cpl.php
#
# =====================================================================
use Error qw(:try);

package LS;

use 5.006;
use strict;
use warnings;

our $VERSION = '1.1.6';

=head1 NAME

LS - Perl module for building clients and servers which resolve and perform
metadata queries on LSIDs.

=head1 SYNOPSIS

 use LS::ID;
 use LS::Locator;

 $lsid = LS::ID->new('URN:LSID:pdb.org:PDB:112L:');

 $locator = LS::Locator->new();
 $authority = $locator->resolveAuthority($lsid);

 $resource = $authority->getResource($lsid);

 $data = $resource->getData();


=head1 DESCRIPTION

The LS module is used for building clients and servers which resolve LSIDs
and perform metadata queries on LSIDs.  More information on LSIDs can be
found at L<http://www.omg.org/cgi-bin/doc?dtc/04-05-01>


=head2 makeAccessorMethods( $methodList, $package )

 Creates default accessor methods for an array ref of method names
 in the package specified by $package

=cut

sub makeAccessorMethods {

	my $methodList = shift;
	my $package = shift;
	
	unless(UNIVERSAL::isa($methodList, 'ARRAY')) {
	
		return undef;
	}
	
	unless($package) {
		
		return undef;
	}
	
	# 
	# Create the accessor / mutator methods for the bindings class
	#
	for my $field (@{ $methodList }) {
	
		no strict "refs";
	
		my $slot = "${package}__${field}";
		my $fn = "${package}::${field}";
		
		*$fn= sub {
	
			my $self = shift;
			
			@_ ? $self->{ $slot } = $_[0] : return $self->{ $slot };
		}
	}
}

=head1 SEE ALSO

L<LS::ID>, L<LS::Locator>, L<LS::Authority>, L<LS::Resource>, L<LS::Service>, L<LS::SOAP::Service>, L<LS::HTTP::Service>,
L<http://www.omg.org/cgi-bin/doc?dtc/04-05-01>, L<http://oss.software.ibm.com/developerworks/projects/lsid>

=head1 AUTHOR

IBM

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2002,2003 IBM Corporation 
All rights reserved.   This program and the accompanying materials
are made available under the terms of the Common Public License v1.0
which accompanies this distribution, and is available at
http://www.opensource.org/licenses/cpl.php

=cut

package LS::Base;

use strict;
use warnings;

use vars qw( $_ERR $_STACK_TRACE );

use Carp qw(:DEFAULT);


=head1 NAME

LS::Base - This package is the base package that all other LS packages inherit from.

=head1 VARIABLES

A list of variables used in this package.

=head2 $_ERR

 Package error string.

=cut

$_ERR = '';

=head2 $_STACK_TRACE

 Holds the Carp stack traces
 
=cut

$_STACK_TRACE = [];

=head2 appendError( $extraMessage )

 Appends additional information to the error string
 Parameters - $extraMessage, Required. The additional information to store.

=cut

sub appendError {

	my ($self, $error) = @_;

	if($self && ref $self) {

		$self->{'_err'} .= "\n -> $error";
	}
	else {

		$_ERR .= "\n -> $error";
	}
}


=head2 recordError( $errorMessage ) - Records the message in the class's error string

 Parameters - $errorMessage, Required. The message to be stored in the class's
 rror string.

=cut

sub recordError {

	my ($self, $error) = @_;

	if($self && ref $self) {

		$self->{'_err'} = $error;
	}
	else {

		$_ERR = $error;
	}
}

=head2 addStackTrace( )

 Appends a Carp stack trace to the error string
 
=cut

sub addStackTrace {

	my $self = shift;
	my $previousTraces = shift;

	if($self && ref $self) {

		# Initialize the stack traces on the first call
		$self->{'_stack_trace'} = [] unless(UNIVERSAL::isa($self->{'_stack_trace'}, 'ARRAY'));

		# Copy an applicable stack traces in before adding our own
		@{ $self->{'_stack_trace'} } = @{ $previousTraces } if( ref $previousTraces eq 'ARRAY');

		push @{ $self->{'_stack_trace'} }, Carp::longmess('Stack trace');
	}
	else {

		# Copy an applicable stack traces in before adding our own
		@{ $_STACK_TRACE } = @{ $previousTraces } if( ref $previousTraces eq 'ARRAY');

		push @{ $_STACK_TRACE }, Carp::longmess('Stack trace');
	}
}

=head2 getStackTrace( )

 Retrieves a copy (in the form of an arrayref) of the 
 stack traces associated with this object.

 Returns - An arrayref that is a COPY of the arrayref containing
 all of the stack traces.

=cut

sub getStackTrace {

	my $self = shift;

	my $copy = [];

	foreach my $st (@{ $self->{'_stack_trace'} }) {

		push @{ $copy }, $st;
	}

	# Eh, there won't be that many of them
	@{ $copy } = reverse( @{ $copy } );

	return $copy;
}


=head2 hasStackTrace( )

 Determins whether or not a stack trace is present

 Returns - True if there is a stack trace
 		   False / undef if no stack trace is available
=cut

sub hasStackTrace {

	my $self = shift;

	if($self && ref $self) {

		return (scalar(@{ $self->{'_stack_trace'} } > 0) );
	}
	else {

		return (scalar(@{ $_STACK_TRACE }) > 0);
	}
}


=head2 clearStackTrace( )

 Clears the stack trace
 
=cut

sub clearStackTrace {

	my $self = shift;

	if($self && ref $self) {

		$self->{'_stack_trace'} = [];
	}
	else {

		$_STACK_TRACE = [];
	}
}


=head2 errorString( )

 Returns one of two error strings:

 1. if the class has been blessed in to a reference, the internal
 class error string is returned.
 2. if the class has not been blessed in to a reference, the package
 error string is returned.

 These values may be undef if no error has occured.
 
=cut

sub errorString {

        my $self = shift;

        if ($self && ref $self) {

		return $self->{'_err'};
        }

        return $_ERR;
}


=head2 error_string( )

 Synonym for errorString
 
=cut

sub error_string {

        my $self = shift;
        return $self->errorString();
}


=head2 errorDetails( )

 Returns the complete error with all applicable 
 stack traces.

 Returns - A very detailed error message
 
=cut

sub errorDetails {

        my $self = shift;

	my $errorMessage;

        if ($self && ref $self) {

		$errorMessage = $self->{'_err'};

		foreach my $st (@{ $self->{'_stack_trace'} }) {

			$errorMessage .= "\n\n" . $st;
		}

		return $errorMessage;
        }

	$errorMessage = $_ERR;

	foreach my $st (@{ $self->{'_stack_trace'} }) {

		$errorMessage .= "\n\n" . $st;
	}

        return $errorMessage;
}

=head1 SEE ALSO

L<LS::ID>, L<LS::Locator>, L<LS::Authority>, L<LS::Resource>, L<LS::Service>, L<LS::SOAP::Service>, L<LS::HTTP::Service>,
L<http://www.omg.org/cgi-bin/doc?dtc/04-05-01>, L<http://oss.software.ibm.com/developerworks/projects/lsid>

=head1 AUTHOR

IBM

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2002,2003 IBM Corporation 
All rights reserved.   This program and the accompanying materials
are made available under the terms of the Common Public License v1.0
which accompanies this distribution, and is available at
http://www.opensource.org/licenses/cpl.php

=cut

package LS::Exception;

use strict;
use warnings;

use base 'Error';
use overload ('""' => 'stringify');


sub new
{
	my $self = shift;
	
	my $text = shift;
	my $object = shift;
	
	local $Error::Depth = $Error::Depth + 1;
	local $Error::Debug = 1;
	
	return $self->SUPER::new(-text=> ($text || ""), -object=> $object);
}


package LS::InvalidParameterException;
use base 'LS::Exception';

package LS::MalformedParameterException;
use base 'LS::Exception';

package LS::RuntimeException;
use base 'LS::Exception';

package LS::ClientException;
use base 'LS::Exception';

1;

__END__

lib/LS/Locator.pm
# ====================================================================
# Copyright (c) 2002,2003 IBM Corporation 
# All rights reserved.   This program and the accompanying materials
# are made available under the terms of the Common Public License v1.0
# which accompanies this distribution, and is available at
# http://www.opensource.org/licenses/cpl.php
# 
# =====================================================================
package LS::Locator;

use strict;
use warnings;

use vars qw( $METHODS );

use Net::DNS::Resolver;

use LS;
use LS::ID;
use LS::Authority;

use base 'LS::Base';


sub BEGIN {

	$METHODS = [
		'authorityCache',
		'localMappingFile',
		'cacheAuthorities',
	];
	
	LS::makeAccessorMethods($METHODS, __PACKAGE__);
}


#
# new( %options ) - Creates a new LS::Locator that searches DNS for an LSID's authority.
#
#	This method accepts parameters in the form of a hash. The following hash keys are
#	used:
#
#		local - Specifies the local filename and path that LS::Locator will use to 
#			short ciruit DNS.
#
#		nocache - If set to true, then DNS queries for authorities will not be cached.
#
#
#	Returns: An LS::Locator object if successful,
#		 undef if there is a failure
#
sub new {
	
	my $self = shift;
	my (%params) = @_;
	
	unless (ref $self) {

		$self = bless {}, $self;

		$self->authorityCache( {} );
	}
	
	my $file = ( $params{'local'} || $params{'localMappingFile'} || 'authorities');
	$self->localMappingFile( $file );
	$self->loadLocalMappingFile();
	$self->cacheAuthorities(1);
	
	$self->cacheAuthorities($params{'nocache'})
		if(exists($params{'nocache'}));
	
	return $self;
}


#
# cacheAuthorities( $bool ) - Returns whether or not Authority queries will be cached.
#
#	 Returns: If Authority queries _WILL_ be cached, 1 is returned
#		  otherwise, a false value is returned 
#


#
# localMappingFile( [$filename] )
# 	Gets/Sets the local filename to use for the static 
#	authority -> endpoint mappings
#
#	Parameters:
#		If the filename parameter is specifed, the local filename is set to that
#		parameter.
#
#	Returns:
#		If the parameter is left unspecified the current local filename  
#		is used.
#


#
# clearCache( ) - Empties the Authority query cache
#
sub clearCache {

	my $self = shift;
	
	$self->authorityCache( {} );

	# Reload the authorities file
	$self->loadLocalMappingFile();
}


#
# clear_cache( ) - Synonym for clearCache
#
sub clear_cache {

	my $self = shift;
	return $self->clearCache();
}


sub loadLocalMappingFile {
	
	my $self = shift;

	# Can only initialize the cache if there is a filename set
	unless($self->localMappingFile()) {
	
		return undef;
	}

	# Load the local authorities file
	unless(open(FILE, $self->localMappingFile()) ) {
	
		return undef;
	}
		
	local ($/)= "\n";
	while (<FILE>) {
		
		# Trim whitespace
		s/^\s+|\s+$//g;

		# Ignore comments
		next if(/^#/);

		my ($name, $location) = split(/\s/, $_, 2);
		next unless($name);

		my ($host, $port, $path) = $location =~ m|^\s*(.*?):(.*?)(/(.*))?$|;
		next unless($host && $port);

		my $authority = LS::Authority->new_by_hostname($host, $port, $path);
		next unless($authority);

		$self->authorityCache()->{ lc($name) } = $authority;
	}

	close(FILE);
}


#
# resolveAuthority( $lsid ) - Resolves the authority for the specified LSID. This method will
#			      search DNS for the hostname and then retrieve the SRV resource 
#			      records for the authority.
#
#	Returns: An LS::Authority object if successful,
#		 undef if an error occurs, check errorString for more information
#
sub resolveAuthority {

	my $self = shift;
	my ($id) = @_;

	my $authority_id;

	unless(UNIVERSAL::isa($id, 'LS::ID')) {

		my $lsid = LS::ID->new($id);
		
		if ($lsid) {
			
			$authority_id = $lsid->authority();
		}
		elsif ($id =~ /^lsidauth:([A-Za-z0-9][\w\(\)\+\,\-\.\=\@\;\$\"\_\!\*\']*)$/i) { 
			$authority_id = $1;
		}
		else {
			
			$self->recordError( 'Cannot resolve invalid LSID: ' . $id );
			$self->addStackTrace() ;
			
			return undef;
		}
	}

	$authority_id = $id->authority();
	unless ($authority_id) {

		$self->recordError( 'Cannot resolve authority, invalid LSID' );
		$self->addStackTrace() ;

		return undef;
	}

	my $cached = $self->authorityCache()->{ $authority_id };

	return $cached 
		if($cached);

	# Use DNS to resolve the authority ID via an SRV record
	my $resolver = Net::DNS::Resolver->new();
	unless($resolver) {

		$self->recordError( 'Unable to create Net::DNS::Resolver object' );
		$self->addStackTrace();

		return undef;
	}

	# Do the DNS lookup
	my $query = $resolver->search('_lsid._tcp.' . $authority_id, 'SRV');
	unless ($query) {

		$self->recordError( 'Unable to resolve authority to SRV record: ' .
				    ($resolver->errorstring() || 'No details available from $resolver') );

		$self->addStackTrace();

		return undef;
	}

	my $rr = ($query->answer())[0];

	# Create an authority interface from the answer
	my $authority = LS::Authority->new_by_hostname($rr->target(), $rr->port());
	unless(UNIVERSAL::isa($authority, 'LS::Authority')) {
	
		$self->recordError('Unable to create new authority from host information');
		$self->addStackTrace();
		
		return undef;
	}

	$self->authorityCache()->{$authority_id} = $authority 
		if($self->cacheAuthorities());

	return $authority;
}
	

#
# resolve_authority - Synonym for resolveAuthority.
#
sub resolve_authority {

	my $self = shift;
	return $self->resolveAuthority(@_);
}




1;


__END__

=head1 NAME

LS::Locator - Resolve authorities for LSIDs

=head1 SYNOPSIS

 use LS::ID;
 use LS::Locator;

 $lsid = LS::ID->new('urn:lsid:ncbi.nlm.nih.gov.lsid.biopathways.org:pubmed:12441807');

 $locator = LS::Locator->new();
 $authority = $locator->resolveAuthority($lsid);

 $auth_host = $authority->host();
 $auth_port = $authority->port();
 $auth_path = $authority->path();

 $resource = $authority->getResource($lsid);
 $data = $resource->getData();

=head1 DESCRIPTION

Objects of the C<LS::Locator> class will resolve the authority of a
given LSID using DNS SRV records. Results are cached for speed.

If a file called F<authorities> is present in the working directory,
the locator's cache is preloaded with the contents of the file. The
authorities file contains a newline separated list of id/location pairs.
Each pair is delimited with whitespace. Lines starting with C<#> are
treated as comments. An example authorities file would look like this:

 # This file enables local resolution of LSID authorities.
 # id      server:port/servicepath

 authority.testing.org   myauthority.org:80/cgi-bin/authority.pl

More information on LSIDs and their resolution can be found at
L<http://www.omg.org/cgi-bin/apps/doc?dtc/04-05-01>

=head1 CONSTRUCTORS

The following method is used to construct a new C<LS::Locator> object:

=over

=item new ( [%options] )

This class method creates a new C<LS::Locator> object and returns it.
Key/value pair arguments may be provided to initialize locator options.
The available options names are C<local> and C<nocache>. The options can
also be set or modified later by method calls described below.

Examples:

 $locator = LS::Locator->new(localMappingFile => 'c:\lsid\authorities.txt');

This creates a new locator object whose cache will be preloaded with the
contents of the file F<c:\lsid\authorities.txt>.

=back

=head1 METHODS

=over

=item loadLocalMappingFile ( )

Loads the entries specified by L<localMappingFile> in to the 
locator's cache. Calling this method will overwrite any 
existing entries in the cache that also appear in the file.

=item localMappingFile ( [$filename] )

Sets or retrieves the name of the local authorities file being used by
the locator. By default, the authorities file is the file called 
F<authorities> in the working directory. To turn off local authority 
resolution, call C<clearCache>, then call this method with an argument 
of C<undef>.

The file is NOT immediately read into the locator's cache, you must
call C<loadLocalMappingFile> which will overwrite any existing entries 
in the cache that also appear in the file.

Examples:

 $locator = LS::Locator->new();
 
 # Load the cache with the contents of f<c:\lsid\myauthorities.txt>
 
 $locator->localMappingFile('c:\lsid\myauthorities.txt');
 $locator->loadLocalMappingFile();
 
 $authority = $locator->resolveAuthority('urn:lsid:ncbi.nlm.nih.gov.lsid.biopathways.org:pubmed:12441807');
 
 # Now turn off local resolution

 $locator->clearCache();
 $locator->localMappingFile(undef);

=item clearCache ( )

Clears the locator's cache of authorities. After the cache is cleared,
it is re-initialized with the contents of the local authorities file, if
present.

=item cacheAuthorities ( $bool )

If this method is called with a true value, the results of future
Authority resolutions will be cached. This does not affect the current
contents of the cache, nor does it affect the reading of a local
authorities file into the cache. It only prevents future DNS respnses
from being added. If the method is called with a defined false value,
e.g. 0, caching of Authority responses will stop. If the method is called
with an undefined argument, the current setting is returned.

=item resolveAuthority ( $id )

Resolves the authority for the given ID. The ID may either be an LSID as
a string or an object of class C<LS::ID>, or an authority ID as a string
of the form lsidauth:authority_id. The return value is an object of
class C<LS::Authority>, or C<undef> if the LSID could not be resolved.
In this case, C<errorString> can be checked for a description of the
error.

Examples:

 $lsid = LS::ID->new('urn:lsid:ncbi.nlm.nih.gov.lsid.biopathways.org:pubmed:12441807');
 $authority = $locator->resolveAuthority($lsid);

 $authority = $locator->resolveAuthority('urn:lsid:ncbi.nlm.nih.gov.lsid.biopathways.org:pubmed:12441807');

 $authority = $locator->resolveAuthority('lsidauth:ncbi.nlm.nih.gov.lsid.biopathways.org');

=item errorString ( )

Returns a description of the last error that occurred in the locator.

Examples:

 $authority = $locator->resolveAuthority('urn:lsid:ncbi.nlm.nih.gov.lsid.biopathways.org:pubmed:12441807');

 if (!$authority) {
 	warn('Unable to resolve authority: ', $locator->errorString(),  '\n');
 }

=back

=head1 COPYRIGHT

Copyright (c) 2002,2003 IBM Corporation.
All rights reserved.   This program and the accompanying materials
are made available under the terms of the Common Public License v1.0
which accompanies this distribution, and is available at
L<http://www.opensource.org/licenses/cpl.php>

=cut


examples/tools/wsdl_parser.pl
#!/usr/bin/perl
# $Id: wsdl_parser.pl,v 1.1 2005/11/11 21:37:26 evanchsa-oss Exp $
# =====================================================================
# Copyright (c) 2005 IBM Corporation 
# All rights reserved.   This program and the accompanying materials
# are made available under the terms of the Common Public License v1.0
# which accompanies this distribution, and is available at
# http://www.opensource.org/licenses/cpl.php
# 
# =====================================================================

use strict;
use warnings;

use LS::Authority::WSDL;
use LS::Authority::WSDL::Simple;

use Data::Dumper;

use Getopt::Long;
use Pod::Usage;

#
# Argument processing
#
my $debug = 0;

my $help = 0;

my @filenames;


my %getOptions = (

	'help'=>	\$help,
	'<>'=>		\&gatherFilenames,
);


pod2usage("$0: Missing command-line options\nTry $0 --help for more information")
	unless(scalar(@ARGV) > 0);

GetOptions( %getOptions ) || pod2usage(2);

pod2usage(-exitstatus=>0, -verbose=> 2) 
	if($help);


#

while(@filenames) {
	my $wsdl_file = shift(@filenames);
	my $wsdl_data = &read_wsdl($wsdl_file);
	
	print "Parsing WSDL file: $wsdl_file\n\n";
	my $wsdl = LS::Authority::WSDL::Simple->from_xml($wsdl_data);
	unless($wsdl) {
		print STDERR "Unable to parse WSDL file\n";
		exit(-1);
	}
	
	&print_wsdl_details($wsdl);
	
	print "\n\n\n";
}

sub read_wsdl {

	my $filename = shift;
	
	my $data;
	unless(open(WSDL, $filename)) {
		print STDERR "$0: Unable to open WSDL file: $filename\n";
		exit(-1);
	}
	local $/ = undef;
	$data = <WSDL>;
	close(WSDL);
	
	return $data;
}

sub gatherFilenames {
	push @filenames, (shift);	
}

sub print_wsdl_details {
	
	my $wsdl = shift;
	
	print "Target Namespace: ", $wsdl->targetNamespace(), "\n\n";
	
	my $i = 0;
	print "Metadata Locations:\n";
	foreach my $metadata_location (@{ $wsdl->getMetadataLocations() }) {
		print "\tMetadata Location $i:\n";
		&print_wsdl_location_details($metadata_location);
		print "\n";
		
		$i++;
	}
	
	if($i == 0) {
		print "\t\tNo metadata locations found\n";
	}

	print "\n\n";
	
	$i = 0;
	print "Data Locations:\n";
	foreach my $data_location (@{ $wsdl->getDataLocations() }) {
		
		print "\tData Location $i:\n";
		&print_wsdl_location_details($data_location);
		print "\n";
		
		$i++;
	}	

	if($i == 0) {
		print "\t\tNo data locations found\n";
	}
}

sub print_wsdl_location_details {
	
	my $location = shift;

	print "\t\tName: ", $location->name(), "\n";
	print "\t\tParent Service Name: ", $location->parentName(), "\n";
	print "\t\tProtocol: ", $location->protocol(), "\n";
	
	if($location->protocol() eq ${LS::Authority::WSDL::Constants::Protocols::HTTP}) {
		print "\t\tHTTP Method: ", $location->method() ,"\n";
	}
	
	print "\t\tURL: ", $location->url(), "\n";
}




__END__

=head1 NAME

wsdl_parser.pl - Parse WSDL containing LSID services and display its details

=head1 SYNOPSIS

    wsdl_parser.pl [options] [ input file(s) ...]

     Options:
       --debug			Turn on debug information within ODO
       --help           	Usage


=head1 OPTIONS

=over 8

=item B<--debug>

Turn on debug information.

=item B<--help>

Print a help message and exits.

=head1 DESCRIPTION

This program will parse a WSDL document looking for LSID service definitions. 
It will then display the details of the components of the LSID service in a
human readable manner.


Debugging information is available with the <--debug> option.

=cut


__END__


Makefile.PL
# Load the Module::Install bundled in ./inc/
    use inc::Module::Install;
    
    # Define metadata
    name            'lsid-perl';
	version 		'1.1.6';
	license 		'cpl';
  	abstract 		'Perl module for building clients and servers which resolve and perform metadata queries on LSIDs.';
    all_from        'lib/LS.pm';
    author 			'IBM <http://lsid.sourceforge.net/>';
	
    # Specific dependencies
    requires        'Carp'              => '0';
    requires        'SOAP::Lite'    	=> '0.69';
	requires        'Compress::Zlib'    => '1.22';
	requires        'URI'    	 		=> '1.21';
	requires        'LWP'    	 		=> '5.53';
	requires        'Net::FTP'    	 	=> '2.65';
	requires        'Net::DNS'    	 	=> '0.28';
	requires        'XML::XPath'    	=> '1.12';
	requires        'MIME::Base64'    	=> '2.12';
	requires        'File::Temp'    	=> '0.12';
	requires        'File::Copy'    	=> '2.01';
	requires        'RDF::Core'    	 	=> '0.30';
	requires        'Time::HiRes'    	=> '1.20';
	requires        'MIME::Tools'    	=> '6.106';
	requires        'Error'    			=> '0.15';
	
	no_index directory => 'examples';
    	
	auto_install;
    WriteAll;